ENH: Holt-Winters simulations

s-scherrer opened the pull request:
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

Implementation of `HoltWintersResults.simulate` as discussed in #6359, as a first step towards prediction intervals.

This is still work in progress, but I would like to get some feedback about the general direction/coding style/implementation details.

I didn't manage to reproduce the results from R's forecast package exactly in the current test, but the difference is low. How exact does it have to be, @ChadFulton?

pep8speaks commented:
Hello @s-scherrer! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/tsa/holtwinters.py`](https://github.com/statsmodels/statsmodels/blob/aed6a666980fbb78d146927f89b299b53f7dd3ba/statsmodels/tsa/holtwinters.py):

> [Line 577:9](https://github.com/statsmodels/statsmodels/blob/aed6a666980fbb78d146927f89b299b53f7dd3ba/statsmodels/tsa/holtwinters.py#L577): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'

s-scherrer commented:
>     * [``](https://github.com/statsmodels/statsmodels/blob/aed6a666980fbb78d146927f89b299b53f7dd3ba/statsmodels/tsa/holtwinters.py):

This is on purpose to follow the mathematical notation in FPP2. Should I change this to somewhat else?

bashtage commented on statsmodels/tsa/holtwinters.py:
```diff
@@ -430,6 +430,227 @@ def _fmt(x):
+        if error == 'add':
+            # estimate error variance from residuals
+            sigma = np.sqrt(np.sum(self.resid**2) / (len(self.resid) - n_params))
+            eps = np.random.randn(steps, nsim) * sigma
```
Need to let a user (optionally) pass in a RandomState rather than using the global generator.

ChadFulton replied:
In addition to supporting RandomState, I think we should also allow a few other cases:

1. User passes in the `eps` they want to use directly (helpful for replication / unit testing)
2. User passes in a scipy distribution function they want to use (allows non-Gaussian errors)
3. (Optional, we can leave this as a TODO) User specifies bootstrapping, in which case we re-sample the in-sample errors.

s-scherrer replied:
I implemented all these with two new keyword arguments:
* `seed` accepts a `RandomState` or an integer
* `random_errors` can be None, an array of values that are used directly, a scipy distribution function, or `"bootstrap"`.

bashtage commented:
You need to check if matplotib is available, and if not skip that part of the test.  Ideally split the matplotlib and non-matpltlib then you can `@pytest.mark.matplotlib` the test that uses MPL to skip it when not available.

bashtage commented on statsmodels/tsa/tests/test_holtwinters.py:
```diff
@@ -358,6 +378,56 @@ def test_hw_seasonal_buggy(self):
+        steps = 20
+        sim = fit.simulate(steps, nsim=nsim, error="mul")
+
+        import matplotlib.pyplot as plt
```
Make a plotting only test.

s-scherrer replied:
Sorry, I should have removed this before. This was just for myself as a first visual sanity check, not a real test. I removed it now.

s-scherrer replied:
Sorry, I forgot to remove this, this was not thought of as a real test, but a first visual sanity check for me. I removed it now.

s-scherrer commented:
I made the following changes:
* Added keyword arguments for passing a random state/seed, a scipy distribution function, fixed errors, or bootstrap
* Moved tests to the end of the test file, using pytest fixtures for the data instead of the test class
* The tests set up a ``HoltWintersResults`` object with exactly the same state as is obtained from R's forecast package for all stable models (the ones that forecast allows). With this, predictions match up to 5 decimals.
* Added a test that checks whether the different keyword options are accepted and run (no comparison of outputs)

ChadFulton commented:
Thanks for this! I should have time for a review this weekend.

s-scherrer commented:
@ChadFulton I just found that there's still an issue when the fit was performed with `use_boxcox=True`. I am working on it.

s-scherrer commented on statsmodels/tsa/holtwinters.py:
```diff
@@ -944,12 +1250,12 @@ def _predict(self, h=None, smoothing_level=None, smoothing_slope=None,
         level = lvls[1:nobs + 1].copy()
         if use_boxcox or use_boxcox == 'log' or isinstance(use_boxcox, float):
             fitted = inv_boxcox(fitted, lamda)
-            level = inv_boxcox(level, lamda)
```
I don't really understand why `level`, `slope` and `season` are "back-transformed" here. For the simulations I need the values in the transformed space, therefore I commented them for now.

The tests are still running fine, does this mean we can remove this?

ChadFulton replied:
That is interesting. I can't see why the back-transformation would be sensible, and I notice that R's forecast package does not back-transform the states.

The only question here is about backwards compatibility - if people are relying on the current behavior. But it looks like the current behavior might be best classified as a bug?

s-scherrer replied:
Yes, I would also classifiy this as a bug. 
What would be the best way to handle this? I could, for a start, store the original values in a private attribute and use these for the simulations. Then we could open a separate issue and PR to discuss backwards compability issues.

ChadFulton replied:
What you suggest is definitely the ideal way forward, but a little annoying for you. If you don't mind doing that, that would be great.

ChadFulton commented on statsmodels/tsa/holtwinters.py:
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
+        b = np.empty((steps + 1, nsim))
+        s = np.empty((steps + m, nsim))
+        # the following uses python's index wrapping
+        lvl[-1,:] = self.level[start]
```
space after comma

ChadFulton commented on statsmodels/tsa/holtwinters.py:
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
+            s[-m:,:] = np.tile(self.season[-m:], (nsim, 1)).T
+        elif 0 <= start and start < m:
+            initial_seasons = self.params["initial_seasons"]
+            _s = np.concatenate((initial_seasons[start+1:],
```
space between operators like `+`

ChadFulton commented:
This is looking great, thanks! I just have a couple of comments.

The only other thing is that the updates should be cleaned up to match up with PEP8, I noted a couple of them (like spaces after commas, or spaces between operators like `+`).

s-scherrer commented on statsmodels/tsa/holtwinters.py:
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
+        ----------
+        steps : int
+            The number of simulation steps.
+        start : int, optional
```
Done, I very liberally copied the documentation and code regarding `anchor` from `MLEModel.simulate`.

ChadFulton commented:
I think this is ready to merge, except for updating the back-transformation to store the original values. If you want to do that, we can merge directly afterwards, assuming CI is still green.

Thanks again!

ChadFulton commented:
This is great, thanks very much for this PR.

s-scherrer commented:
Thanks a lot for your assistance. I had a lot of fun and learned quite a bit.
